File: updraft-sdk/src/main/java/com/appswithlove/updraft/manager/CheckFeedbackEnabledManager.java
package com.appswithlove.updraft.manager;

import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.LifecycleObserver;
import androidx.lifecycle.OnLifecycleEvent;
import androidx.lifecycle.ProcessLifecycleOwner;

import com.appswithlove.updraft.interactor.CheckFeedbackEnabledInteractor;
import com.appswithlove.updraft.interactor.CheckFeedbackResultModel;
import com.appswithlove.updraft.presentation.UpdraftSdkUi;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

public class CheckFeedbackEnabledManager implements LifecycleObserver, ShakeDetectorManager.ShakeDetectorListener {

    private UpdraftSdkUi mUpdraftSdkUi;
    private CheckFeedbackEnabledInteractor mCheckFeedbackEnabledInteractor;

    private Disposable mCheckFeedbackDisposable;

    public CheckFeedbackEnabledManager(UpdraftSdkUi updraftSdkUi,
                                       CheckFeedbackEnabledInteractor checkFeedbackEnabledInteractor) {
        mUpdraftSdkUi = updraftSdkUi;
        mCheckFeedbackEnabledInteractor = checkFeedbackEnabledInteractor;
    }

    public void start() {
        ProcessLifecycleOwner.get().getLifecycle().addObserver(this);
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_START)
    public void onAppForegrounded() {
        runFeedbackCheck();
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_STOP)
    public void onAppBackgrounded() {
        if (mCheckFeedbackDisposable != null) {
            mCheckFeedbackDisposable.dispose();
        }
    }

    private void runFeedbackCheck() {
        if (mCheckFeedbackDisposable != null) {
            mCheckFeedbackDisposable.dispose();
        }
        mCheckFeedbackDisposable = mCheckFeedbackEnabledInteractor
                .run()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        checkFeedbackResultModel -> {
                            if (!checkFeedbackResultModel.isFeedbackEnabled() && mUpdraftSdkUi.isCurrentlyShowingFeedback()) {
                                mUpdraftSdkUi.closeFeedback();
                                return;
                            }
                            if (checkFeedbackResultModel.isShowAlert()) {
                                if (checkFeedbackResultModel.getAlertType() == CheckFeedbackResultModel.ALERT_TYPE_FEEDBACK_DISABLED) {
                                    mUpdraftSdkUi.showFeedbackDisabledAlert();
                                }
                                if (checkFeedbackResultModel.getAlertType() == CheckFeedbackResultModel.ALERT_TYPE_HOW_TO_GIVE_FEEDBACK) {
                                    mUpdraftSdkUi.showHowToGiveFeedbackAlert();
                                }
                            }
                        },
                        Throwable::printStackTrace
                );
    }

    @Override
    public void onShakeDetected() {
        mUpdraftSdkUi.showFeedback();
        runFeedbackCheck();
    }
}

File: updraft-sdk/src/main/java/com/appswithlove/updraft/manager/CheckFeedbackEnabledManager.kt
package com.appswithlove.updraft.manager

import androidx.lifecycle.DefaultLifecycleObserver
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.ProcessLifecycleOwner
import androidx.lifecycle.lifecycleScope
import com.appswithlove.updraft.interactor.CheckFeedbackEnabledInteractor
import com.appswithlove.updraft.interactor.CheckFeedbackResultModel
import com.appswithlove.updraft.presentation.UpdraftSdkUi
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch

class CheckFeedbackEnabledManager(
    private val updraftSdkUi: UpdraftSdkUi,
    private val checkFeedbackEnabledInteractor: CheckFeedbackEnabledInteractor
) : DefaultLifecycleObserver, ShakeDetectorManager.ShakeDetectorListener {

    private var checkFeedbackJob: Job? = null

    fun start() {
        ProcessLifecycleOwner.get().lifecycle.addObserver(this)
    }

    override fun onStart(owner: LifecycleOwner) {
        super.onStart(owner)
        runFeedbackCheck(owner)
    }

    override fun onStop(owner: LifecycleOwner) {
        super.onStop(owner)
        checkFeedbackJob?.cancel()
    }

    private fun runFeedbackCheck(owner: LifecycleOwner) {
        checkFeedbackJob?.cancel()

        checkFeedbackJob = owner.lifecycleScope.launch {
            try {
                val checkFeedbackResultModel = checkFeedbackEnabledInteractor.run()

                if (!checkFeedbackResultModel.isFeedbackEnabled && updraftSdkUi.isCurrentlyShowingFeedback) {
                    updraftSdkUi.closeFeedback()
                    return@launch
                }

                if (checkFeedbackResultModel.showAlert) {
                    when (checkFeedbackResultModel.alertType) {
                        CheckFeedbackResultModel.ALERT_TYPE_FEEDBACK_DISABLED ->
                            updraftSdkUi.showFeedbackDisabledAlert()

                        CheckFeedbackResultModel.ALERT_TYPE_HOW_TO_GIVE_FEEDBACK ->
                            updraftSdkUi.showHowToGiveFeedbackAlert()
                    }
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    override fun onShakeDetected() {
        updraftSdkUi.showFeedback()

        // Run feedback check again in a global lifecycle scope since Shake might happen outside lifecycle event
        ProcessLifecycleOwner.get().lifecycleScope.launch {
            try {
                val checkFeedbackResultModel = checkFeedbackEnabledInteractor.run()

                if (!checkFeedbackResultModel.isFeedbackEnabled && updraftSdkUi.isCurrentlyShowingFeedback) {
                    updraftSdkUi.closeFeedback()
                    return@launch
                }

                if (checkFeedbackResultModel.showAlert) {
                    when (checkFeedbackResultModel.alertType) {
                        CheckFeedbackResultModel.ALERT_TYPE_FEEDBACK_DISABLED ->
                            updraftSdkUi.showFeedbackDisabledAlert()

                        CheckFeedbackResultModel.ALERT_TYPE_HOW_TO_GIVE_FEEDBACK ->
                            updraftSdkUi.showHowToGiveFeedbackAlert()
                    }
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }
}
